Be much more explicit about variable precedence as it applies to roles (#73939)
ansible/ansible: docs/docsite/rst/user_guide/playbooks_variables.rst

  #. role dependency defaults for tasks from a role (defined in ``roles/{dependent_roles}/defaults/main.yml``)
  #. role defaults for tasks from a role (defined in ``role/defaults/main.yml``) [1]_ [8]_ [9]_
  #. inventory file or script group vars [2]_
  #. inventory ``group_vars/all`` [3]_
  #. playbook ``group_vars/all`` [3]_
  #. inventory ``group_vars/*`` [3]_
  #. playbook ``group_vars/*`` [3]_
  #. inventory file or script host vars [2]_
  #. inventory ``host_vars/*`` [3]_
  #. playbook ``host_vars/*`` [3]_
  #. host facts / cached ``set_facts`` [4]_
  #. play ``vars``
  #. play ``vars_prompt``
  #. play ``vars_files``
  #. role vars from all roles (defined in ``roles/*/vars/main.yml`` or ``vars:``)
  #. role dependency vars for tasks from a role (defined in ``roles/{dependent_roles}/vars/main.yml``)
  #. role vars for tasks from a role (defined in ``role/vars/main.yml``) [5]_ [8]_ [9]_
  #. block ``vars`` (only for tasks in block)
  #. task ``vars`` (only for the task)
  #. ``include_vars``
  #. ``set_facts`` / registered vars
  #. role dependency params for tasks from a role
  #. role params for tasks from a role [6]_ [7]_

>   #. include params

Author: Try to indicate that these are task include params, and not role includes, since role includes do not support params.

Ansible does apply variable precedence, and you might have a use for it. Here is the order of precedence from least to greatest (the last listed variables override all other variables):

  #. command line values (for example, ``-u my_user``, these are not variables)
  #. role defaults from all roles (defined in ``roles/*/defaults/main.yml``)
  #. role dependency defaults for tasks from a role (defined in ``roles/{dependent_roles}/defaults/main.yml``)
  #. role defaults for tasks from a role (defined in ``role/defaults/main.yml``) [1]_ [8]_ [9]_
  #. inventory file or script group vars [2]_
  #. inventory file relative ``group_vars/all`` [3]_
  #. playbook file relative ``group_vars/all`` [3]_
  #. inventory file relative ``group_vars/*`` [3]_

>   #. playbook file relative ``group_vars/*`` [3]_

Reviewer: Might read better reversed like:

~~~rst
#. ``group_vars/all`` relative to inventory file [3]_
#. ``group_vars/all`` relative to playbook file [3]_
#. ``group_vars/*`` relative to inventory file [3]_
#. ``group_vars/*`` relative to playbook file [3]_
~~~

  #. inventory file or script host vars [2]_
  #. inventory file relative ``host_vars/*`` [3]_

>   #. playbook file relative ``host_vars/*`` [3]_

Reviewer: Might read better reversed like:

~~~rst
#. ``host_vars/*`` relative to inventory file [3]_
#. ``host_vars/*`` relative to playbook file [3]_
~~~

  #. host facts / cached ``set_facts`` [4]_
  #. play ``vars``
  #. play ``vars_prompt``
  #. play ``vars_files``

>   #. role vars from all roles (defined in ``roles/*/vars/main.yml`` or ``vars:``)

Reviewer 2: Should `vars` have a colon here, but not for similar cases in lines above (`play vars`, `play vars_prompt`, `play vars_files`)?

Reviewer: I think the colon is useful to visually cue the reader that it should be defined in an ansible/playbook/task/whatever file rather than in a standalone vars file.

However, as a reader of the docs, I would not know where this `vars:` would be defined. Where would you put `vars:` inside a role? On a task? Does `vars:` on a task count as a roles var here? Does this mean if I have a list of roles defined in a playbook, and set a variable for one role, the other roles get that variable too?

  #. role dependency vars for tasks from a role (defined in ``roles/{dependent_roles}/vars/main.yml``)
  #. role vars for tasks from a role (defined in ``role/vars/main.yml``) [5]_ [8]_ [9]_

>   #. block ``vars`` (only for tasks in block)

Reviewer: missing `a` between "in" and "block".

Reviewer: or make `block` plural

.. rubric:: Footnotes

.. [1] Tasks in each role see their own role's defaults if specified, or those from the role's dependencies. Tasks defined outside of a role see the last role's defaults.
.. [2] Variables defined in inventory file or provided by dynamic inventory.
.. [3] Includes vars added by 'vars plugins' as well as host_vars and group_vars which are added by the default vars plugin shipped with Ansible.

> .. [4] When created with set_facts's cacheable option, variables have the high precedence in the play, but are the same as a host facts precedence when they come from the cache.

Reviewer 3: I don't understand this note yet. Is this suggestion still accurate?
```suggestion
.. [4] Variables created with set_facts where cacheable=yes have higher precedence in the play than host facts. Variables created with set_facts without cacheable=yes have the same precedence as host facts.
```
Also, how can variables from two different sources have the same precedence . . . if they are different, which one wins?

Author: fwiw, I didn't change anything with the wording here, I just removed the line wrap.  Basically `set_facts` without `cacheable=yes` create and overwrite host facts, as they exist in the same "dict".

Reviewer 3: I understand this a bit better after our conversation today. I'll close this suggestion and possibly put in a slightly different one.

> .. [4] When created with set_facts's cacheable option, variables have the high precedence in the play, but are the same as a host facts precedence when they come from the cache.

Reviewer 3: ```suggestion
.. [4] Variables created with set_facts where cacheable=yes have higher precedence in the play than host facts. Variables created with set_facts where cacheable=no (the default setting) are treated as host facts. If a variable created with set_facts where cacheable=no supplies a different value than a pre-existing host fact, the set_facts value will override the pre-existing host fact.
```

.. [5] Tasks in each role see their own role's vars if specified, or those from the role's dependencies. Tasks defined outside of a role see the last role's vars.

> .. [6] Tasks in each role see their own role's params if specified, or those from the role's dependencies. Role params are not exposed outside of the role.

Reviewer 3: This feels a bit like mixing apples and oranges. Are `role params` really **variables**, or are they a different object, similar to playbook keywords? If they're not variables, then maybe we could describe them as `role keywords` instead of `role params`, and/or include this information in https://docs.ansible.com/ansible/latest/reference_appendices/general_precedence.html#general-precedence-rules and create better links to that page . . .

Author: role params are a specially scoped version of vars.

Reviewer 2: role params (why not parameters?):
```yml
roles:
    - role: role_one
      parameter_1: 123
    - name: role_two # this is an alias for role, so never use a parameter "name" or "role"
      parameter_1: 234
```

role vars:
from `roles/x/vars/main.yml` and ...
```yaml
roles:
    - role: role_one
      vars:
        var_1: 123
    - name: role_two
      vars:
        name: val_of_var_name # no conflict
tasks:
  - import_role:
      name: role_3
    vars:
      var1: 345
  - include_role:
      name: role_4
    vars:
      var1: 456
```

`params` and `vars` are used the same way in the role. They only have differences in the scope. (And some of this scope difference is probably unexpected, see #47606 and #36274.)

> .. [7] Role params are nto specified within ``vars:`` they are top level params specified as part of the role definition

Author: ```suggestion
.. [7] Role params are not specified within ``vars:`` they are top level params specified as part of the role definition
```

> .. [8] Roles loaded using ``include_role`` do not have their defaults or vars available in the play scope until they are encountered during the playbook execution, and when ``public: true`` is defined. Use of ``roles:`` or ``import_role`` makes the defaults and vars available to the play scope at playbook parsing time.

Reviewer 3: This is useful information, could we add a sentence or note about what difference this makes for the user? Do role vars for tasks in a role hold a different "rank" in variable precedence when the role is loaded with `include_role` because those vars haven't been loaded yet when the main comparison is made? If so, is the precedence rank lower or higher than when the role is loaded with `import_role` or just in a `roles:` list?

Author: I feel like you just broke my brain.  They precedence doesn't change, but due to the way includes work, a variable may resolve differently before and after the include.

> .. [9] Roles do not share a parent relationship, and as such, all roles have their parent set to the play. There is no special handling of a role that includes or imports another role from a variable precedence perspective. A role that is included or imported from another role exists in the play scope, and does not use special handling of the vars/defaults of the role that included it. In the specific case described here, the role that includes another, is not considered a dependency.

Reviewer 3: ```suggestion
.. [9] Even when one role calls another role, they do not share a parent-child relationship. All roles have their parent set to the play. There is no special handling of a role that includes or imports another role from a variable precedence perspective. A role that is included or imported from another role exists in the play scope, and does not use special handling of the vars/defaults of the role that included it. In the specific case described here, the role that includes another, is not considered a dependency.
```
